src/main/java/org/commcare/formplayer/application/MenuController.java

```diff
@@ -163,8 +166,14 @@ public EntityDetailListResponse getDetails(@RequestBody SessionNavigationBean se
     @UserRestore
     @AppInstall
     public BaseResponseBean navigateSessionWithAuth(@RequestBody SessionNavigationBean sessionNavigationBean,
-            @CookieValue(Constants.POSTGRES_DJANGO_SESSION_ID) String authToken,
-            HttpServletRequest request) throws Exception {
+                        @CookieValue(Constants.POSTGRES_DJANGO_SESSION_ID) String authToken,
+                        @CookieValue(value = FORMPLAYER_SESSION_LANGUAGE, required = false) String sessionPrefLang,
+                        HttpServletRequest request,
+                        HttpServletResponse response) throws Exception {
+
+        if (sessionPrefLang == null || !sessionPrefLang.equals(sessionNavigationBean.getLocale())) {
+            response.addCookie(new Cookie(FORMPLAYER_SESSION_LANGUAGE, sessionNavigationBean.getLocale()));
+        }
```

Comment:
I don't follow what's going on here. We're resetting the cookie value if it doesn't match what's being sent in the request body. But the value in the request body is coming from the cookie right (unless the cookie is empty in which case the first part of the condition is met)?